src/query.jl

```diff
@@ -0,0 +1,257 @@
+### Format registry infrastructure
+
+@doc """
+`DataFormat{sym}()` indicates a known binary or text format of kind `sym`,
+where `sym` is always a symbol. For example, a .csv file might have
+`DataFormat{:CSV}()`.
+
+An easy way to write `DataFormat{:CSV}` is `format"CSV"`.
+""" ->
+immutable DataFormat{sym} end
+
+macro format_str(s)
+    :(DataFormat{$(Expr(:quote, symbol(s)))})
+end
+
+const unknown_df = DataFormat{:UNKNOWN}
+
+@doc """
+`unknown(f)` returns true if the format of `f` is unknown.""" ->
+unknown(::Type{format"UNKNOWN"}) = true
+unknown{sym}(::DataFormat{sym}) = unknown(sym)
+unknown(::Any) = false
+
+const ext2sym = Dict{ASCIIString,Union(Symbol,Vector{Symbol})}()
+const magic_list = Array(Pair, 0)    # sorted, see magic_cmp below
+const sym2info = Dict{Symbol,Any}()  # Symbol=>(magic, extension)
+const magic_func = Array(Pair, 0)    # for formats with complex magic #s
+
+@doc """
+`add_format(fmt, magic, extention)` registers a new `DataFormat`.
```

Comment:
Good point. I'll do that, and at the same time merge all my README tweaks into the original commit.